core/src/test/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/chained/ChainedChangeMoveTest.java

```diff
@@ -11,176 +11,145 @@
 import ai.timefold.solver.core.impl.domain.variable.descriptor.GenuineVariableDescriptor;
 import ai.timefold.solver.core.impl.domain.variable.inverserelation.SingletonInverseVariableSupply;
 import ai.timefold.solver.core.impl.heuristic.selector.SelectorTestUtils;
+import ai.timefold.solver.core.impl.move.director.MoveDirector;
 import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
 import ai.timefold.solver.core.impl.testdata.domain.chained.TestdataChainedAnchor;
 import ai.timefold.solver.core.impl.testdata.domain.chained.TestdataChainedEntity;
 import ai.timefold.solver.core.impl.testdata.domain.chained.TestdataChainedSolution;
 import ai.timefold.solver.core.impl.testdata.util.PlannerTestUtils;
 
 import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
 
 class ChainedChangeMoveTest {
 
+    private final GenuineVariableDescriptor<TestdataChainedSolution> variableDescriptor = TestdataChainedEntity
+            .buildVariableDescriptorForChainedObject();
+    private final InnerScoreDirector<TestdataChainedSolution, SimpleScore> innerScoreDirector =
+            PlannerTestUtils.mockScoreDirector(variableDescriptor.getEntityDescriptor().getSolutionDescriptor());
+
     @Test
     void noTrailing() {
-        GenuineVariableDescriptor<TestdataChainedSolution> variableDescriptor = TestdataChainedEntity
-                .buildVariableDescriptorForChainedObject();
-        InnerScoreDirector<TestdataChainedSolution, SimpleScore> scoreDirector =
-                PlannerTestUtils.mockScoreDirector(variableDescriptor.getEntityDescriptor().getSolutionDescriptor());
-
-        TestdataChainedAnchor a0 = new TestdataChainedAnchor("a0");
-        TestdataChainedEntity a1 = new TestdataChainedEntity("a1", a0);
-        TestdataChainedEntity a2 = new TestdataChainedEntity("a2", a1);
-        TestdataChainedEntity a3 = new TestdataChainedEntity("a3", a2);
+        var a0 = new TestdataChainedAnchor("a0");
+        var a1 = new TestdataChainedEntity("a1", a0);
+        var a2 = new TestdataChainedEntity("a2", a1);
+        var a3 = new TestdataChainedEntity("a3", a2);
 
-        TestdataChainedAnchor b0 = new TestdataChainedAnchor("b0");
-        TestdataChainedEntity b1 = new TestdataChainedEntity("b1", b0);
+        var b0 = new TestdataChainedAnchor("b0");
+        var b1 = new TestdataChainedEntity("b1", b0);
 
-        SingletonInverseVariableSupply inverseVariableSupply = SelectorTestUtils.mockSingletonInverseVariableSupply(
+        var inverseVariableSupply = SelectorTestUtils.mockSingletonInverseVariableSupply(
                 new TestdataChainedEntity[] { a1, a2, a3, b1 });
 
-        ChainedChangeMove<TestdataChainedSolution> move =
-                new ChainedChangeMove<>(variableDescriptor, a3, b1, inverseVariableSupply);
-        assertThat(move.isMoveDoable(scoreDirector)).isTrue();
-        ChainedChangeMove<TestdataChainedSolution> undoMove = move.createUndoMove(scoreDirector);
-        move.doMove(scoreDirector);
```

Comment:
This comment applies to all places where we are removing the undo validation. I understand the move logic has changed, but I am still concerned about removing the undo validation entirely.

I wonder how much effort it would take to use the ephemeral director while maintaining the undo validations.

Reply:
In my opinion, there is no need. All undo operations are now centralized on the recording score director and the actions, so all testing belongs there. Doing it on the moves would be duplication.

In some tests, I still do it, because those tests rely on the state resulting from the undo moves. So there, the undo is necessary for the test to be correct.